File: rich_game_mechnism/src/com/tw/core/Player.java
package com.tw.core;

import com.tw.core.commands.Command;
import com.tw.core.places.Place;
import com.tw.core.responses.Response;
import com.tw.core.tools.Gift;

/**
 * Created by pzzheng on 11/16/16.
 */
public class Player {
    protected Status status;
    private int stuckDays;
    private Game game;
    private Place currentPlace;
    private Command lastCommand;
    private Asest asests;
    private int luckyDays;

    private Player() {
        status = Status.WAIT_FOR_TURN;
        asests = new Asest();
    }

    public Status getStatus() {
        return status;
    }

    public Status execute(Command command) {
        if (status.equals(Status.WAIT_FOR_COMMAND)) {
            lastCommand = command;
            status = command.execute(this);
        }
        return status;
    }

    public Status inTurn() {
        status = Status.WAIT_FOR_COMMAND;
        return status;
    }

    public Status endTurn() {
        status = Status.WAIT_FOR_TURN;
        if (luckyDays > 0) luckyDays--;
        if (stuckDays > 0) stuckDays--;
        return status;
    }

    public Status waitForResponse() {
        status = Status.WAIT_FOR_RESPONSE;
        return status;
    }

    public Status bankrupt() {
        status = Status.BANKRUPT;
        return status;
    }

    public Status respond(Response response) {
        if (status.equals(Status.WAIT_FOR_RESPONSE)) {
            status = lastCommand.respond(response, this);
        }
        return status;
    }

    public boolean isStucked() {
        return stuckDays > 0;
    }

    public static Player createPlayerWithGame_Fund_CommandState(Game game, int initialFund) {
        Player player = new Player();
        player.game = game;
        player.asests.addFund(initialFund);
        player.status = Status.WAIT_FOR_COMMAND;
        return player;
    }

    public Place getCurrentPlace() {
        return currentPlace;
    }

    public Game getGame() {
        return game;
    }

    public Command lastCommand() {
        return lastCommand;
    }

    public void setLastCommand(Command lastCommand) {
        this.lastCommand = lastCommand;
    }

    public Asest getAsests() {
        return asests;
    }

    public void moveTo(Place place) {
        currentPlace = place;
    }

    public void getLuckyGod() {
        luckyDays = Gift.LUCKY_GOD.getValue() + 1;
    }

    public boolean isLucky() {
        return luckyDays > 0;
    }

    public void stuckFor(int stuckDays) {
        this.stuckDays = stuckDays + 1;
    }

    public enum Status {WAIT_FOR_COMMAND, BANKRUPT, WAIT_FOR_RESPONSE, WAIT_FOR_TURN}
}

File: ruby/rich_game_ruby/src/player.rb
require_relative 'asset'
require_relative '../src/assistencePower/gift'
class Player
  attr_reader :game, :asset, :currentPlace
  attr_accessor :lastResponsiveCommand, :status


  def initialize(game, initialFund=10000)
    @game = game
    @asset = Asset.new(initialFund)
    @status = Status::WAIT_FOR_TURN
    @luckyDays = 0
    @stuckDays = 0
  end

  def moveTo(place)
    @currentPlace = place
  end

  def self.create_player_with_game_and_fund(game, initialFund = 10000)
    return Player.new(game, initialFund)
  end

  def self.create_player_with_game_and_fund_and_command_state(game, initialFund=10000)
    player = Player.new(game, initialFund)
    player.status = Status::WAIT_FOR_COMMAND
    return player
  end

  def execute(command)
    @status=@status.action(command, self)
  end

  def waitForResponse
    @status = Status::WAIT_FOR_RESPONSE
    return @status
  end

  def waitForCommand
    @status = Status::WAIT_FOR_COMMAND
    return @status
  end

  def inTurn
    @status = Status::WAIT_FOR_COMMAND
    return @status
  end

  def endTurn
    @luckyDays -= 1 if(@luckyDays > 0)
    @stuckDays -= 1 if(@stuckDays > 0)
    @status = Status::WAIT_FOR_TURN
    game.nextPlayer
    return @status
  end

  def bankrupt
    @status = Status::BANKRUPT
    return @status
  end

  def getLuckyGod
    @luckyDays = Gift::LUCKY_GOD.value + 1
  end

  def isLucky
    @luckyDays > 0
  end

  def stuckFor(days)
    @stuckDays = days + 1
  end

  def isStuck
    @stuckDays > 0
  end

  class Status
    WAIT_FOR_COMMAND = Status.new

    def WAIT_FOR_COMMAND.action(command, player)
      command.execute(player)
    end

    WAIT_FOR_RESPONSE = Status.new

    def WAIT_FOR_RESPONSE.action(command, player)
      return player.lastResponsiveCommand.respond(command, player)
    end

    WAIT_FOR_TURN = Status.new

    BANKRUPT = Status.new

    def action(command, player)
    end

  end

end